src/main/cpp/src/case_when.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/mr/device/per_device_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace spark_rapids_jni {
+
+/**
+ *
+ * Select the column index for the first true in bool columns.
+ * For the row does not contain true, use end index(number of columns).
+ *
+ * e.g.:
+ *   column 0 in table: true,  false, false, false
+ *   column 1 in table: false, true,  false, false
+ *   column 2 in table: false, false, true, false
+ *
+ *   1st row is: true, flase, false; first true index is 0
+ *   2nd row is: false, true, false; first true index is 1
+ *   3rd row is: false, flase, false; first true index is 2
+ *   4th row is: false, false, false; do not find true, set index to the end index 3
+ *
+ *   output column: 0, 1, 2, 3
+ *   In the `case when` context, here 3 index means using NULL value.
+ *
+ */
+std::unique_ptr<cudf::column> select_first_true_index(
+  cudf::table_view const& when_bool_columns,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
+
+/**
+ *
+ * Select strings in scalar column according to index column.
+ * If index is out of bound, use NULL value
+ * e.g.:
+ *   scalar column: s0, s1, s2
+ *   index  column: 0,  1,  2,  2,  1,  0,  3
+ *   output column: s0, s1, s2, s2, s1, s0, NULL
+ *
+ */
+std::unique_ptr<cudf::column> select_from_index(
+  cudf::strings_column_view const& then_and_else_scalar_column,
+  cudf::column_view const& select_index_column,
+  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

Comment:
This is a very good optimization strategy and it should not be limited to string type. Indeed, we can apply this to any data type. Thus, the function signature should take the input as a regular `column_view`, not `string_column_view`. Supporting other data types is not required to be implemented in this PR but the function signature should be prepared for future changes.
```suggestion
 * Select strings in scalar column according to index column.
 * If index is out of bound, use NULL value
 * e.g.:
 *   scalar column: s0, s1, s2
 *   index  column: 0,  1,  2,  2,  1,  0,  3
 *   output column: s0, s1, s2, s2, s1, s0, NULL
 *
 */
std::unique_ptr<cudf::column> select_from_index(
  cudf::column_view const& case_when_options,
  cudf::column_view const& select_index_column,
  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```

Reply:
Oh wait. I just realize that this is just a gather. So we don't need this function at all. Just call `cudf::gather`, which already supports all data types.